core/src/test/java/com/google/errorprone/bugpatterns/threadsafety/ImmutableCheckerTest.java

```diff
@@ -2993,4 +2993,62 @@ public void switchExpressionsResultingInGenericTypes_doesNotThrow() {
             "}")
         .doTest();
   }
+
+  @Test
+  public void switchExpressionsYieldStatement_doesNotThrow() {
+    assumeTrue(RuntimeVersion.isAtLeast14());
+    compilationHelper
+        .addSourceLines(
+            "Test.java",
+            "import java.util.function.Supplier;",
+            "class Test {",
+            "  String test(String mode) {",
+            "    return switch (mode) {",
+            "      case \"random\" -> {",
+            "        yield \"foo\";",
+            "      }",
+            "      default -> throw new IllegalArgumentException();",
+            "    };",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void switchExpressionsMethodReference_doesNotThrow() {
+    assumeTrue(RuntimeVersion.isAtLeast14());
+    compilationHelper
+        .addSourceLines(
+            "Test.java",
+            "import java.util.function.Supplier;",
+            "class Test {",
+            "  Supplier<Double> test(String mode) {",
+            "    return switch (mode) {",
+            "      case \"random\" -> Math::random;",
+            "      default -> throw new IllegalArgumentException();",
+            "    };",
+            "  }",
+            "}")
+        .doTest();
+  }
```

Comment:
These two cases are irrelevant for this change, but I added to make sure NPE is only related to `yield AClass::methodReference`.